# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/windows/filedialog/WinFileDialogParent.cpp
# Commit: af562cec3d70
# Full Hash: af562cec3d706ee3ed605ffeade4d3ba982d2545
# Author: Ray Kraesig <rkraesig@mozilla.com>
# Date: 2023-10-27 04:08:13
# Regressor Bug: 1837079
# File Overlap Count: 2
# Description:
#   Bug 1837079 - [5/10] Create WinFileDialog IPDL protocol and implementation  r=gstoll,handyman,ipc-reviewers,nika,win-reviewers,mhowell
#   
#   Create and implement a new top-level IPC protocol, `PWinFileDialog`,
#   using the primitives from bug 1833450 and the new sandboxing type from
#   the previous commit.
# ==============================================================================

diff -r b5d52e5f75f2 -r af562cec3d70 widget/windows/filedialog/WinFileDialogParent.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/widget/windows/filedialog/WinFileDialogParent.cpp	Thu Oct 26 18:21:28 2023 +0000
@@ -0,0 +1,51 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "mozilla/widget/filedialog/WinFileDialogParent.h"
+
+#include "mozilla/Logging.h"
+#include "mozilla/Result.h"
+#include "mozilla/SpinEventLoopUntil.h"
+#include "nsISupports.h"
+
+mozilla::LazyLogModule sLogWFD("FileDialog");
+
+namespace mozilla::widget::filedialog {
+
+WinFileDialogParent::WinFileDialogParent() {
+  MOZ_LOG(sLogWFD, LogLevel::Info, ("%s %p", __PRETTY_FUNCTION__, this));
+}
+
+WinFileDialogParent::~WinFileDialogParent() {
+  MOZ_LOG(sLogWFD, LogLevel::Info, ("%s %p", __PRETTY_FUNCTION__, this));
+}
+
+PWinFileDialogParent::nsresult WinFileDialogParent::BindToUtilityProcess(
+    mozilla::ipc::UtilityProcessParent* aUtilityParent) {
+  Endpoint<PWinFileDialogParent> parentEnd;
+  Endpoint<PWinFileDialogChild> childEnd;
+  nsresult rv = PWinFileDialog::CreateEndpoints(base::GetCurrentProcId(),
+                                                aUtilityParent->OtherPid(),
+                                                &parentEnd, &childEnd);
+
+  if (NS_FAILED(rv)) {
+    MOZ_ASSERT(false, "Protocol endpoints failure");
+    return NS_ERROR_FAILURE;
+  }
+
+  if (!aUtilityParent->SendStartWinFileDialogService(std::move(childEnd))) {
+    MOZ_ASSERT(false, "SendStartWinFileDialogService failed");
+    return NS_ERROR_FAILURE;
+  }
+
+  if (!parentEnd.Bind(this)) {
+    MOZ_ASSERT(false, "parentEnd.Bind failed");
+    return NS_ERROR_FAILURE;
+  }
+
+  return NS_OK;
+}
+}  // namespace mozilla::widget::filedialog